packages/block-library/src/search/edit.js

```diff
@@ -446,33 +447,34 @@ export default function SearchEdit( {
 							value={ `${ width }${ widthUnit }` }
 							units={ units }
 						/>
-						<ButtonGroup
-							className="wp-block-search__components-button-group" // unused, kept for backwards compatibility
-							aria-label={ __( 'Percentage Width' ) }
+						<ToggleGroupControl
+							label={ __( 'Percentage Width' ) }
+							value={
+								PERCENTAGE_WIDTHS.includes( width ) &&
+								widthUnit === '%'
+									? width
+									: undefined
+							}
```

Comment:
Passing the value without a condition results in a "too many re-renders" error. Should I add an inline note?

Reply:
I'll look into it in case this is something that we should solve, rather than avoid

Reply:
Yes. This definitely looks like a bug.

Reply:
So! The issue seems to be because `width` can become `NaN` when unit control's value is empty, and I think that's because of a bug in the `onChange` handler of `UnitControl`, which doesn't handle non-number strings.

Applying this diff seems to fix the issue:

```diff
diff --git a/packages/block-library/src/search/edit.js b/packages/block-library/src/search/edit.js
index 7e7cd855a9..fa8256baf2 100644
--- a/packages/block-library/src/search/edit.js
+++ b/packages/block-library/src/search/edit.js
@@ -425,13 +425,11 @@ export default function SearchEdit( {
 							}
 							step={ 1 }
 							onChange={ ( newWidth ) => {
-								const filteredWidth =
-									widthUnit === '%' &&
-									parseInt( newWidth, 10 ) > 100
-										? 100
-										: newWidth;
+								const parsedNewWidth = parseInt( newWidth, 10 );
 								setAttributes( {
-									width: parseInt( filteredWidth, 10 ),
+									width: Number.isNaN( parsedNewWidth )
+										? undefined
+										: parsedNewWidth,
 								} );
 							} }
 							onUnitChange={ ( newUnit ) => {
@@ -449,12 +447,7 @@ export default function SearchEdit( {
 						/>
 						<ToggleGroupControl
 							label={ __( 'Percentage Width' ) }
-							value={
-								PERCENTAGE_WIDTHS.includes( width ) &&
-								widthUnit === '%'
-									? width
-									: undefined
-							}
+							value={ width }
 							hideLabelFromVision
 							onChange={ ( newWidth ) => {
 								setAttributes( {

```

(I don't think there's a need to clamp the percentage to `100` in the `onChange` function, because we're already using the `min` and `max` props)

This seems to also fix another potential bug that can be experienced on `trunk`:
- setting a width (either via the `ButtonGroup` or `UnitControl`)
- delete the numeric value in `UnitControl`
- notice how nothing happens (since the `width` is internally set to `NaN`)

With the changes suggested above, the `width` value should be successfully updated to `undefined`, which (I believe) should then implicitly apply whatever is the default value?

Reply:
I actually extracted most of the diff (plus another little fix) to a separate PR: https://github.com/WordPress/gutenberg/pull/65468

Once that's merged, we should be able to rebase and just set `value={ width }` here

Reply:
https://github.com/WordPress/gutenberg/pull/65468 is now merged

Reply:
Updated in [475a7e1](https://github.com/WordPress/gutenberg/pull/65340/commits/475a7e19472d50e999a19b1504b09b6cb82ba947).